patches/server/0966-Paper-Item-API.patch

```diff
@@ -0,0 +1,5048 @@
++
++        return patterns.remove(i);
++    }
++
++    @Override
++    public void setPattern(int i, @NotNull Pattern pattern) {
++        PropertyHolder.Mutable holder = this.propertyHolder.getOrCreate(ItemProperties.BLOCK_ENTITY_TAG);
++        List<Pattern> patterns = holder.getOrCreate(PaperMetaProperties.PATTERNS);
++
++        patterns.set(i, pattern);
++    }
++
++    @Override
++    public int numberOfPatterns() {
++        PropertyHolder.Mutable holder = this.propertyHolder.getOrCreate(ItemProperties.BLOCK_ENTITY_TAG);
++        List<Pattern> patterns = holder.getOrCreate(PaperMetaProperties.PATTERNS);
++
++        return patterns.size();
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/item/meta/MetaBlockState.java b/src/main/java/io/papermc/paper/item/meta/MetaBlockState.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..87f5aac48685d91ca55d4753ca1a417d52d278e3
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/item/meta/MetaBlockState.java
+@@ -0,0 +1,103 @@
++package io.papermc.paper.item.meta;
++
++import com.destroystokyo.paper.MaterialSetTag;
++import io.papermc.paper.inventory.item.properties.ItemProperties;
++import io.papermc.paper.property.PropertyHolder;
++import io.papermc.paper.property.nbt.AdapterContext;
++import io.papermc.paper.property.nbt.NbtPropertyHolder;
++import org.apache.commons.lang.Validate;
++import org.bukkit.DyeColor;
++import org.bukkit.Material;
++import org.bukkit.block.BlockState;
++import org.bukkit.craftbukkit.block.CraftBanner;
++import org.bukkit.craftbukkit.block.CraftBlockEntityState;
++import org.bukkit.craftbukkit.block.CraftBlockStates;
++import org.bukkit.inventory.meta.BlockStateMeta;
++import org.jetbrains.annotations.NotNull;
++
++public class MetaBlockState extends PaperMeta implements BlockStateMeta {
++
++    public MetaBlockState(PropertyHolder.Mutable mutable, Material material) {
++        super(mutable, material);
++    }
++
++    @Override
++    public boolean hasBlockState() {
++        return this.propertyHolder.hasProperty(ItemProperties.BLOCK_ENTITY_TAG);
++    }
++
++    @Override
++    public BlockState getBlockState() {
++        PropertyHolder.Mutable holder = this.propertyHolder.get(ItemProperties.BLOCK_ENTITY_TAG);
++
++        Material stateMaterial = (this.material != Material.SHIELD) ? this.material : shieldToBannerHack(holder); // Only actually used for jigsaws
++        if (holder != null) {
++            if (this.material == Material.SHIELD) {
++                holder.set(PaperMetaProperties.IDENTIFIER, "minecraft:banner");
++            } else if (this.material == Material.BEE_NEST || this.material == Material.BEEHIVE) {
++                holder.set(PaperMetaProperties.IDENTIFIER, "minecraft:beehive");
++            } else if (MaterialSetTag.SHULKER_BOXES.isTagged(this.material)) {
++                holder.set(PaperMetaProperties.IDENTIFIER, "minecraft:shulker_box");
++            }
```

Comment:
As far as I can tell, these fixes are no longer needed. Vanilla always includes the block entity id in the itemstack nbt data now (as of a 1.19 snapshot)